torch/csrc/distributed/c10d/Utils.cpp

```diff
@@ -160,13 +171,18 @@ std::pair<int, PortType> listen(PortType port) {
   struct ::addrinfo hints, *res = NULL;
   std::memset(&hints, 0x00, sizeof(hints));
   hints.ai_flags = AI_PASSIVE | AI_ADDRCONFIG;
-  hints.ai_family = AF_SELECTED; // IPv4 on Windows, IPv4/6 on Linux
+  hints.ai_family = AF_INET6; // default to IPv6
```

Comment:
Another technique would be to specify `AF_UNSPEC` which requests `getaddinfo()` to return both IPv4 and IPv6 addresses and then do the filtering in the `while` loop on line 196 which I believe would be cleaner:

```
// Semi-pseudo code that prioritizes IPv6 over IPv4.
for (auto ai_family : [AF_INET6, AF_INET]) {
   while (nextAddr) {
        if (nextAddr->ai_family != ai_family) {
           continue;
        }
        ...
}
```

Reply:
That is similar to what specifying `/etc/gai.conf` would achieve, but without the side-effects.
That wouldn't work on Windows though, as it doesn't have `AF_UNSPEC`. This should work anywhere.

Reply:
Another thing that I noticed after googling a bit is apparently on Windows the default for `IPV6_V6ONLY` is 1, meaning binding to a IPv6 socket on Windows (and if for some reason an admin changed it to 1 in sysctl on Linux) will disable IPv4 connections. I believe to keep BC we should also make sure that we call `setsockopt()` and explicitly set `IPV6_V6ONLY` to 0.